### DISCUSSION TITLE
Using conformance witnesses as individual values

### DISCUSSION THREAD

**Discussion by kyouko-taiga** [hylo-lang/hylo#1084] [Q&A]
In Scala, one defines trait conformance by creating instances of a type class. Consider this example:

In Hylo:
```
trait BitSet {
  fun is_set(at position: Int) -> Bool
}

extension BitSet {
  public fun some_generic_algorithm() -> Bool { is_set(at: 1) }
}

conformance Int64: BitSet {
  public fun is_set(at position: Int) -> Bool { (self & (1 << position)) != 0 }
}

public fun main() {
  print((31 as Int64).some_generic_algorithm())
}
```

In Scala:
```scala
trait BitSet[T] {
  extension (self: T)
    def isSet(position: Int): Boolean
}

extension [T](self: T)(using BitSet[T])
  def someGenericAlgorithm = self.isSet(1)

given BitSet[Long] with
  extension (self: Long) def isSet(position: Int) = (self & (1 << position)) != 0

println((31L).someGenericAlgorithm)
```
Let's review important differences:

`BitSet` is called a "type class" in Scala circles. AFAIK it is the same concept as what we call a trait in Hylo (it's even introduced with the same keyword!), though Scala enthusiasts seem to associate traits with inheritance.

Instances of type classes can be understood as witnesses, or proofs, of a type's conformance to a trait. In Hylo, we create these witnesses with conformance declarations. In Scala, we instantiate type classes. In the example, that's done by declaring an implicit anonymous value with `given`.

In Hylo, we can write generic algorithms in extensions of traits, or by bounding generic parameters by traits:
```
// BitSet.some_generic_algorithm as a free function
fun some_generic_algorithm<T: BitSet>(self: T) -> Bool { (self & (1 << position)) != 0 }
```
In Scala, the witness of `T`'s conformance to `BitSet` is just another parameter. In the example, it is declared implicit, meaning that Scala will use whatever is in scope when we call `someGenericAlgorithm`.

One advantage of Scala's approach is that we can treat instances of type classes (i.e., witnesses) as individual values. So we can very naturally instantiate one in a scope and let it be used somewhere else. All we need to do is to give it a name:
```scala
def dynamicWitness() : BitSet[Long] = {
  given w: BitSet[Long] with ...
  w
}

// The type class returned by `dynamicWitness` is used to call `someGenericAlgorithm`
given BitSet[T] = dynamicWitness
println((31L).someGenericAlgorithm)
```

In Hylo, the witness of `Int64 : BitSet` is `Int64` itself and there is no way to make it escape. We can create an existential and package the conformances we want that are in scope, but that requires creating an instance. For example, `Int64.zero() as any BitSet where Self == Int64` is roughly equivalent to `(0L, BitSet[Long])` in Scala.

There are two situations in which this feature may be useful:
1. We want to apply a generic algorithm on `T` but override the latter's conformance(s). For example, given `T : Comparable`, we want to change the meaning of `T.infix<` in the context of a single call to `Array<T>.min`.
2. The implementation of `T : C` requires information read from a different type `U`. For example, `T` might represent indices in a collection and telling whether an instance is lesser than another requires access to that collection. *(note for Swift users: whether or not `Collection.Index` should be `Comparable` in Hylo is an orthogonal discussion.)*

We can partially address (1) using scoped conformance, which probably covers a lot of useful cases. Also note that it lets us apply static dispatch whereas, AFAICT, every call to a method defined by a type class is dynamic in Scala. Scoped conformance do not really address dynamic use of conformances, but maybe that's okay.

I think (2) is an important use case. We have floated the idea of using implicit contexts (can't find where 😞) but haven't landed on anything concrete yet. One challenging issue is that if a conformance needs to capture an instance of another type `U`, then either witnesses have to be dynamic or all requirements must implicitly take an additional `U`, which has to be thread through the entire call stack. That seems difficult even with implicit parameters.

**Comment by tothambrus11** [comment]
Update: we have adopted a similar design. See [The State of Generics](https://github.com/hylo-lang/hylo-new/blob/main/Docs/State-of-Generics.md) for more information.
